File: 2022/elixir/day5/answer.exs
file = Path.expand('2022/input/day5/example') |> Path.absname()
input = File.read!(file)

[state, instruction] =
  input |> String.split("\n\n", trim: true) |> Enum.map(&String.split(&1, "\n", trim: true))

defmodule Day5 do
  def parse_state(state) do
    state
    |> Enum.map(fn line ->
      String.codepoints(line)
      |> Enum.chunk_every(4)
      |> Enum.map(&Enum.join/1)
      |> Enum.map(&String.trim/1)
    end)
    |> Enum.reverse()
    |> build_stack()
  end

  def build_stack(state) do
    [head | lines] = state

    lines
    |> Enum.reduce([], fn (line, acc) ->
      line
      |> Enum.with_index()
      |> Enum.each(fn {c, i} ->
        if Enum.at(acc, i) do
          acc =
        else

        end
      end)
      end
    end)
  end
end

Day5.parse_state(state)

File: 2022/javascript/day5/answer.js
const fs = require("fs")

const input = fs.readFileSync("../../input/day5/input", { encoding: "utf8", flag: "r" })

const [puzzle, instruction] = input.split('\n\n')

const data = puzzle.split('\n')
data.pop()
const instructions = instruction.split('\n').filter(i => i)

function chunk(input, chunkSize) {
    const arr = []
    for (let i = 0; i < input.length; i += chunkSize) {
        arr.push(input.slice(i, i + chunkSize).trim().replace(/(\[|\])/g, ''))
    }
    return arr
}

const stacks = data.reverse().reduce((acc, row) => {
    chunk(row, 4).forEach((c, idx) => {
        if (c) {
            if (acc[idx]) {
                acc[idx].push(c)
            } else {
                acc[idx] = [c]
            }
        }
    })
    return acc
}, [])

const stacks2 = data.reduce((acc, row) => {
    chunk(row, 4).forEach((c, idx) => {
        if (c) {
            if (acc[idx]) {
                acc[idx].push(c)
            } else {
                acc[idx] = [c]
            }
        }
    })
    return acc
}, [])

function run(stacks, instructions) {
    instructions.forEach(instruction => {
        const [_, count, from, to] = instruction.match(/move ([0-9]+) from ([0-9]+) to ([0-9]+)/)
        stacks[to - 1].push(...stacks[from - 1].splice(stacks[from - 1].length - count).reverse())
    })
}

function run2(stacks, instructions) {
    instructions.forEach(instruction => {
        const [_, count, from, to] = instruction.match(/move ([0-9]+) from ([0-9]+) to ([0-9]+)/)
        stacks[to - 1].push(...stacks[from - 1].splice(stacks[from - 1].length - count))
    })
}

function getLast(arr) {
    return arr.reduce((str, row) => {
        return str += row[row.length - 1]
    }, '')
}

run(stacks, instructions)
const answer1 = getLast(stacks)
console.log(answer1)

run2(stacks2, instructions)
const answer2 = getLast(stacks2)
console.log(answer2)
